src/libpsl-native/src/createprocess.cpp

```diff
@@ -0,0 +1,225 @@
+    {
+        while (CheckInterrupted(result = fcntl(pipeFds[0], F_SETFD, FD_CLOEXEC)));
+        if (result == 0)
+        {
+            while (CheckInterrupted(result = fcntl(pipeFds[1], F_SETFD, FD_CLOEXEC)));
+        }
+
+        if (result != 0)
+        {
+            int tmpErrno = errno;
+            close(pipeFds[0]);
+            close(pipeFds[1]);
+            errno = tmpErrno;
+        }
+    }
+
+    return result;
+}
+
+int32_t ForkAndExecProcess(
+    const char* filename,
+    char* const argv[],
+    char* const envp[],
+    const char* cwd,
+    int32_t redirectStdin,
+    int32_t redirectStdout,
+    int32_t redirectStderr,
+    int32_t creationFlags,
+    int32_t* childPid,
+    int32_t* stdinFd,
+    int32_t* stdoutFd,
+    int32_t* stderrFd)
+{
+    int success = true;
+    int processId = -1;
+    int stdinFds[2] = { -1, -1 };
+    int stdoutFds[2] = { -1, -1 };
+    int stderrFds[2] = { -1, -1 };
+
+    // Validate arguments
+    if (nullptr == filename || nullptr == argv || nullptr == envp || nullptr == stdinFd || nullptr == stdoutFd ||
+        nullptr == stderrFd || nullptr == childPid)
+    {
+        assert(false && "null argument.");
+        errno = EINVAL;
+        success = false;
+        goto done;
+    }
+
+    if ((redirectStdin & ~1) != 0 || (redirectStdout & ~1) != 0 || (redirectStderr & ~1) != 0)
+    {
+        assert(false && "Boolean redirect* inputs must be 0 or 1.");
+        errno = EINVAL;
+        success = false;
+        goto done;
+    }
+
+    // Make sure we can find and access the executable. exec will do this, of course, but at that point it's already
+    // in the child process, at which point it'll translate to the child process' exit code rather than to failing
+    // the Start itself.  There's a race condition here, in that this could change prior to exec's checks, but there's
+    // little we can do about that. There are also more rigorous checks exec does, such as validating the executable
+    // format of the target; such errors will emerge via the child process' exit code.
+    if (access(filename, X_OK) != 0)
+    {
+        success = false;
+        goto done;
+    }
+
+    // Open pipes for any requests to redirect stdin/stdout/stderr
+    if ((redirectStdin && SystemNative_Pipe(stdinFds, O_CLOEXEC) != 0) || 
+        (redirectStdout && SystemNative_Pipe(stdoutFds, O_CLOEXEC) != 0) ||
+        (redirectStderr && SystemNative_Pipe(stderrFds, O_CLOEXEC) != 0))
+    {
+        success = false;
+        goto done;
+    }
+
+    // Fork the child process
+    if ((processId = fork()) == -1)
+    {
+        success = false;
+        goto done;
+    }
+
+    if (processId == 0) // processId == 0 if this is child process
+    {
+        // Close the child's copy of the parent end of any open pipes
+        CloseIfOpen(stdinFds[WRITE_END_OF_PIPE]);
+        CloseIfOpen(stdoutFds[READ_END_OF_PIPE]);
+        CloseIfOpen(stderrFds[READ_END_OF_PIPE]);

```

Comment:
There are some additional changes for the close-on-exec -- calls to `CloseIfOpen` are removed from the `if (processId == 0)` block, see the CoreFx code [here](https://github.com/dotnet/corefx/blob/release/2.0.0/src/Native/Unix/System.Native/pal_process.cpp#L167).

I may be easier to see the differences in the PR change: https://github.com/dotnet/corefx/commit/6d204a6a83e3eca283535ca930157350d4adabd6